### PR TITLE
Support checking for a REST API request before the query is parsed

### PR DESCRIPTION
**Pull request by TimothyBJacobs**
Trac ticket: https://core.trac.wordpress.org/ticket/42061

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/load.php [hunk 1]
```diff
@@ -1490,6 +1490,96 @@ function wp_doing_cron() {
 	return apply_filters( 'wp_doing_cron', defined( 'DOING_CRON' ) && DOING_CRON );
 }
 
+/**
+ * Determines whether the current request is a REST API request.
```
Comment (peterwilsoncc) [inline src/wp-includes/load.php:1494]: CS Nit

```suggestion
 * Determine whether the current request is a REST API request.
```

File: src/wp-includes/load.php [hunk 2]
```diff
@@ -1490,6 +1490,96 @@ function wp_doing_cron() {
...
+function wp_doing_rest() {
+	if ( defined( 'REST_REQUEST' ) || did_action( 'parse_request' ) ) {
+		$doing_rest = defined( 'REST_REQUEST' ) && REST_REQUEST;
+
+		/** This filter is documented in wp-includes/load.php */
+		return apply_filters( 'wp_doing_rest', $doing_rest );
+	}
+
+	// This intentionally doesn't use wp_using_themes(). We are trying to check if wp() would be called for this request
+	// in a front-end context which is typically indicated by the WP_USE_THEMES constant.
```
Comment (peterwilsoncc) [inline src/wp-includes/load.php:1512]:
```
/*
 * CS: multiline
 * comment format
 */
```
🔢

File: src/wp-includes/load.php [hunk 3]
```diff
@@ -1490,6 +1490,96 @@ function wp_doing_cron() {
...
+		return apply_filters( 'wp_doing_rest', $doing_rest );
+	}
+
+	if ( isset( $GLOBALS['wp_rewrite'] ) ) {
+		$index = $GLOBALS['wp_rewrite']->index;
+	} else {
+		$index = 'index.php';
+	}
+
+	$pathinfo         = isset( $_SERVER['PATH_INFO'] ) ? $_SERVER['PATH_INFO'] : '';
```
Comment (peterwilsoncc) [inline src/wp-includes/load.php:1526]: I know this is a mostly a copy-paste from `WP` but a couple of tiny things could be tidied up along the way.

```suggestion
	$path_info         = isset( $_SERVER['PATH_INFO'] ) ? $_SERVER['PATH_INFO'] : '';
```
:1234:

File: src/wp-includes/load.php [hunk 4]
```diff
@@ -1490,6 +1490,96 @@ function wp_doing_cron() {
...
+	list( $pathinfo ) = explode( '?', $pathinfo );
+	$pathinfo         = str_replace( '%', '%25', $pathinfo );
+
+	list( $req_uri ) = explode( '?', $_SERVER['REQUEST_URI'] );
+	$home_path       = trim( parse_url( home_url(), PHP_URL_PATH ), '/' );
```
Comment (peterwilsoncc) [inline src/wp-includes/load.php:1531]: Core helper if it's defined when this runs. 

```suggestion
	$home_path       = trim( wp_parse_url( home_url(), PHP_URL_PATH ), '/' );
```

**Review by peterwilsoncc — COMMENTED** [review]
I've added a couple of comments to this, having some trouble working out how to test locally so I be sure about the parsing. As it comes from the main WP class I assume it's pretty well battle tested though.

File: src/wp-includes/load.php (hunk 4 above)
Comment (pbiron) [reply src/wp-includes/load.php:1531]: If the return value of `home_url()` does not contain a path (e.g. "https://example.com") then in PHP 8.1+ this results in a "trim(): Passing null to parameter #1 ($string) of type string is deprecated" warning.

Thus, this should probably be replaced with something like:

```
$home_url_path   = wp_parse_url( home_url(), PHP_URL_PATH );
$home_path       = $home_url_path ? trim( $home_url_path, '/' ) : '';
```

**Comment by felixarntz** [comment]
While the purpose of this PR is different, in terms of https://core.trac.wordpress.org/ticket/42061 it is now superseded by #5658.

We may want to open another ticket for the specific point of checking for a REST API request _prior_ to parsing the query, as that's a more specific use-case than what was mostly discussed on the ticket.

**Pull request closed without merging**